### ISSUE TITLE
XLA produces wrong results when a Keras Model.call decorated with @tf.function is invoked from an outer tf.function(jit_compile=True)

### ISSUE THREAD

**Issue by tinywisdom** [tensorflow/tensorflow#118314]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

0.21.0

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04.4 LTS (x86_64)

### Mobile device

Ubuntu 22.04.4 LTS (x86_64)

### Python version

3.10.16

### Current behavior

When a Keras `Model.call` method is itself decorated with `@tf.function`, invoking the model from an outer `tf.function(jit_compile=True)` produces a numerically incorrect result.

The same model invoked directly and from an outer non-XLA `tf.function(jit_compile=False)` produces identical results. A control model with the same `Dense` layer but without `@tf.function` on `call()` is stable under XLA, with only normal floating-point-level differences.

This appears to be a silent wrong-result issue specific to the combination:
```python
@tf.function
def Model.call(...):
    ...

@tf.function(jit_compile=True)
def outer(...):
    return model(...)
```
The issue reproduces on Host XLA, without using a logical GPU.

### Standalone code to reproduce the issue

```shell
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import json
import os
import platform

import numpy as np
import tensorflow as tf


SEED = 0
BATCH = 32
IN_DIM = 16
OUT_DIM = 10
RTOL = 1e-5
ATOL = 1e-6


def banner(title):
    print("\n" + "=" * 100)
    print(title)
    print("=" * 100)


def summarize_tensor(name, x, n=8):
    arr = x.numpy()
    print(
        f"{name}: shape={arr.shape}, dtype={arr.dtype}, "
        f"first_values={arr.reshape(-1)[:n].tolist()}"
    )


def compare(name, a, b):
    a_np = a.numpy()
    b_np = b.numpy()

    same_shape = a_np.shape == b_np.shape
    if same_shape:
        abs_diff = np.abs(a_np.astype(np.float64) - b_np.astype(np.float64))
        denom = np.maximum(
            np.maximum(np.abs(a_np.astype(np.float64)), np.abs(b_np.astype(np.float64))),
            1e-12,
        )
        rel_diff = abs_diff / denom
        max_abs = float(np.max(abs_diff)) if abs_diff.size else 0.0
        max_rel = float(np.max(rel_diff)) if rel_diff.size else 0.0
        allclose = bool(np.allclose(a_np, b_np, rtol=RTOL, atol=ATOL))
    else:
        max_abs = None
        max_rel = None
        allclose = False

    result = {
        "same_shape": same_shape,
        "allclose": allclose,
        "shape_a": list(a_np.shape),
        "shape_b": list(b_np.shape),
        "dtype_a": str(a_np.dtype),
        "dtype_b": str(b_np.dtype),
        "max_abs_diff": max_abs,
        "max_rel_diff": max_rel,
        "preview": [],
    }

    flat_a = a_np.reshape(-1)
    flat_b = b_np.reshape(-1)
    for i in range(min(8, flat_a.size, flat_b.size)):
        result["preview"].append(
            {
                "i": i,
                "a": float(flat_a[i]),
                "b": float(flat_b[i]),
                "abs": float(abs(float(flat_a[i]) - float(flat_b[i]))),
            }
        )

    print(f"\nCompare {name}:")
    print(json.dumps(result, indent=2, sort_keys=True))
    return result


class CleanDense(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.dense = tf.keras.layers.Dense(OUT_DIM, use_bias=True)

    def call(self, x, training=False):
        return self.dense(x)


class NestedDense(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.dense = tf.keras.layers.Dense(OUT_DIM, use_bias=True)

    @tf.function
    def call(self, x, training=False):
        return self.dense(x)


def make_input():
    values = np.linspace(-1.0, 1.0, BATCH * IN_DIM, dtype=np.float32)
    return tf.constant(values.reshape(BATCH, IN_DIM), dtype=tf.float32)


def make_weights():
    kernel = np.linspace(
        -0.07, 0.07, IN_DIM * OUT_DIM, dtype=np.float32
    ).reshape(IN_DIM, OUT_DIM)
    bias = np.linspace(-0.03, 0.03, OUT_DIM, dtype=np.float32)
    return kernel, bias


def build_and_assign(model, x):
    # First call creates variables.
    _ = model(x, training=False)

    kernel, bias = make_weights()
    model.dense.kernel.assign(kernel)
    model.dense.bias.assign(bias)

    return model


def run_case(case_name, model, x):
    banner(case_name)

    direct_1 = model(x, training=False)
    direct_2 = model(x, training=False)

    @tf.function(jit_compile=False)
    def graph_fn(inp):
        return model(inp, training=False)

    @tf.function(jit_compile=True)
    def xla_fn(inp):
        return model(inp, training=False)

    graph_out = graph_fn(x)
    xla_out = xla_fn(x)

    summarize_tensor("direct_1", direct_1)
    summarize_tensor("direct_2", direct_2)
    summarize_tensor("graph_out", graph_out)
    summarize_tensor("xla_out", xla_out)

    c_repeat = compare("direct_1 vs direct_2", direct_1, direct_2)
    c_graph = compare("direct_1 vs graph_out", direct_1, graph_out)
    c_xla = compare("direct_1 vs xla_out", direct_1, xla_out)
    c_graph_xla = compare("graph_out vs xla_out", graph_out, xla_out)

    return {
        "repeat": c_repeat,
        "direct_vs_graph": c_graph,
        "direct_vs_xla": c_xla,
        "graph_vs_xla": c_graph_xla,
    }


def main():
    np.random.seed(SEED)
    tf.random.set_seed(SEED)

    # Force Host XLA by hiding logical GPUs.
    try:
        tf.config.set_visible_devices([], "GPU")
    except RuntimeError:
        pass

    banner("Environment")
    print("python:", platform.python_version())
    print("platform:", platform.platform())
    print("tensorflow:", tf.__version__)
    print("CUDA_VISIBLE_DEVICES:", os.environ.get("CUDA_VISIBLE_DEVICES", ""))
    print("physical GPUs:", [str(x) for x in tf.config.list_physical_devices("GPU")])
    print("logical GPUs:", [str(x) for x in tf.config.list_logical_devices("GPU")])
    print("run_functions_eagerly:", tf.config.functions_run_eagerly())

    x = make_input()

    banner("Input")
    summarize_tensor("x", x)
    print("input min:", float(tf.reduce_min(x).numpy()))
    print("input max:", float(tf.reduce_max(x).numpy()))
    print("input mean:", float(tf.reduce_mean(x).numpy()))

    clean_model = build_and_assign(CleanDense(), x)
    nested_model = build_and_assign(NestedDense(), x)

    clean = run_case("CASE A: CleanDense, plain call()", clean_model, x)
    nested = run_case("CASE B: NestedDense, @tf.function on call()", nested_model, x)

    banner("FINAL_STATUS")

    status = {
        "clean_direct_vs_xla_allclose": clean["direct_vs_xla"]["allclose"],
        "clean_direct_vs_xla_max_abs_diff": clean["direct_vs_xla"]["max_abs_diff"],
        "nested_direct_vs_graph_allclose": nested["direct_vs_graph"]["allclose"],
        "nested_direct_vs_xla_allclose": nested["direct_vs_xla"]["allclose"],
        "nested_direct_vs_xla_max_abs_diff": nested["direct_vs_xla"]["max_abs_diff"],
        "nested_graph_vs_xla_allclose": nested["graph_vs_xla"]["allclose"],
        "nested_graph_vs_xla_max_abs_diff": nested["graph_vs_xla"]["max_abs_diff"],
    }

    if (
        clean["direct_vs_xla"]["allclose"]
        and nested["direct_vs_graph"]["allclose"]
        and not nested["direct_vs_xla"]["allclose"]
    ):
        status["status"] = "xla_wrong_result_candidate"
        status["interpretation"] = (
            "CleanDense is stable under XLA, and NestedDense is stable under non-XLA graph, "
            "but NestedDense produces a different result under outer tf.function(jit_compile=True)."
        )
    elif not clean["direct_vs_xla"]["allclose"]:
        status["status"] = "not_minimal_nested_tf_function_specific"
        status["interpretation"] = (
            "CleanDense also mismatches under XLA, so this is not specific to nested @tf.function."
        )
    else:
        status["status"] = "not_reproduced"
        status["interpretation"] = (
            "The nested @tf.function XLA mismatch was not reproduced by this minimal Dense case."
        )

    print("FINAL_STATUS=" + json.dumps(status, sort_keys=True))


if __name__ == "__main__":
    main()
```

### Relevant log output

```shell
# The control model without @tf.function on call() behaves correctly under XLA:

CASE A: CleanDense, plain call()

direct_1 vs direct_2:
  allclose: true
  max_abs_diff: 0.0

direct_1 vs graph_out:
  allclose: true
  max_abs_diff: 0.0

direct_1 vs xla_out:
  allclose: true
  max_abs_diff: 5.21540641784668e-08

# However, the model whose call() is decorated with @tf.function produces a wrong result under the outer XLA-compiled function:

CASE B: NestedDense, @tf.function on call()

direct_1 vs direct_2:
  allclose: true
  max_abs_diff: 0.0

direct_1 vs graph_out:
  allclose: true
  max_abs_diff: 0.0

direct_1 vs xla_out:
  allclose: false
  max_abs_diff: 3.9610073566436768
  max_rel_diff: 1.999407449577972

graph_out vs xla_out:
  allclose: false
  max_abs_diff: 3.9610073566436768
  max_rel_diff: 1.999407449577972

# The first few values also differ substantially:

direct_1:
[1.8397265672683716, 0.09311047196388245, 2.305582046508789,
 -3.2128374576568604, -0.8142527341842651, -0.1703890562057495,
 0.47825542092323303, 0.892669141292572]

xla_out:
[0.037312328815460205, -0.7209519147872925, -1.6554253101348877,
 0.3475034236907959, 0.6117182970046997, 0.07731732726097107,
 0.1041094958782196, -0.8910849094390869]

# The final summary from the reproducer is:

FINAL_STATUS={
  "clean_direct_vs_xla_allclose": true,
  "clean_direct_vs_xla_max_abs_diff": 5.21540641784668e-08,
  "nested_direct_vs_graph_allclose": true,
  "nested_direct_vs_xla_allclose": false,
  "nested_direct_vs_xla_max_abs_diff": 3.9610073566436768,
  "nested_graph_vs_xla_allclose": false,
  "nested_graph_vs_xla_max_abs_diff": 3.9610073566436768,
  "status": "xla_wrong_result_candidate"
}
```

**Comment by bharatjetti** [comment]
Hi @tinywisdom , Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.

**Comment by MRiffiAslett** [comment]
**Target**: comment on existing open issue https://github.com/tensorflow/tensorflow/issues/118314 (NOT a new issue — this looks like the same underlying bug)

---

I ran into what looks like the same issue from a different angle, via a TensorFlow/XLA fuzzing project (WhiteFox-based), and it doesn't need the nested inner-`@tf.function`-plus-outer-`jit_compile=True` shape from the original report. Decorating `Model.call()` directly with `@tf.function(jit_compile=True)` and calling it once — with the layer's variables built lazily inside that same call, rather than pre-built — reproduces the same kind of silent wrong result:

```python
import random
import numpy as np
import tensorflow as tf

x = tf.constant([[[-1.0, 0.0], [-2.0, -3.0]], [[4.0, 5.0], [6.0, 7.0]]])

# `exec()` here isn't cosmetic: hand-written equivalents that skip it (e.g.
# defining two named model classes directly, or wrapping the class def in a
# plain function) reliably drew *different* Dense-kernel values and were not
# reproducible run-to-run for me -- Keras's layer-initializer seeding seems
# sensitive to it. This exact shape reproduced the same numbers deterministically
# across three separate process runs.
test_code = '''
class MyModel(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.layer = tf.keras.layers.Dense(1, activation=tf.keras.activations.tanh)
    def call(self, inputs):
        return self.layer(inputs)

m = MyModel()
out = m(x)
'''

def run(jit_compile):
    random.seed(42); np.random.seed(42); tf.random.set_seed(42)
    code = test_code if not jit_compile else test_code.replace(
        "    def call", "    @tf.function(jit_compile=True)\n    def call"
    )
    g = {"tf": tf, "x": x}
    exec(code, g)
    return g["out"], g["m"]

out_naive, m_naive = run(jit_compile=False)
out_xla, m_xla = run(jit_compile=True)  # first (and only) call -- layer builds inside the jit-compiled call

print("naive :", out_naive.numpy().flatten())
print("xla   :", out_xla.numpy().flatten())
print("naive kernel:", m_naive.layer.get_weights()[0].flatten())
print("xla   kernel:", m_xla.layer.get_weights()[0].flatten())
```

naive (eager, undecorated `call`, same seed):
```
naive : [-0.76704603 -0.99780214  0.999994    1.        ]
```
`@tf.function(jit_compile=True)` directly on `call()`, first (and only) invocation:
```
xla   : [-0.21142544  0.78863823 -0.9269943  -0.97599864]
```
Autocluster (`TF_XLA_FLAGS=--tf_xla_auto_jit=2`, no decorator on `call()`) matches eager exactly. `get_weights()` reads back numerically identical kernel/bias across all three modes — the divergence is in what the compiled program computes, not in which weights it's using.

This reproduced deterministically (identical output and identical readback weights) across four independent fresh-process runs. FWIW it also showed up unprompted across 20 structurally different fuzzer-generated models (different layer counts/types — `Dense`, `Conv2D`+`BatchNormalization`, etc.) in the same run, all sharing this same "layer builds inside a `jit_compile=True`-decorated `call()`" shape, which suggests the trigger condition might be broader than the nested-`tf.function` case in the original report. Happy to share the fuzzer output / more of those cases if useful for narrowing down the root cause.

Environment: TensorFlow 2.20.0-dev0 (self-built from source, CPU-only), Python 3.12.3, Ubuntu 24.04.4 LTS.
